src/Microsoft.Extensions.Configuration.Abstractions/ConfigurationExtensions.cs

```diff
@@ -26,17 +26,28 @@ public static string GetConnectionString(this IConfiguration configuration, stri
         /// </summary>
         /// <param name="configuration">The <see cref="IConfiguration"/> to enumerate.</param>
         /// <returns>An enumeration of key value pairs.</returns>
-        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration)
+        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration) => configuration.AsEnumerable(removePathFromChildKeys: false);
+
+        /// <summary>
+        /// Get the enumeration of key value pairs within the <see cref="IConfiguration" />
+        /// </summary>
+        /// <param name="configuration">The <see cref="IConfiguration"/> to enumerate.</param>
+        /// <param name="removePathFromChildKeys">If true, removes the configuration path from the child keys.</param>
+        /// <returns>An enumeration of key value pairs.</returns>
+        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration, bool removePathFromChildKeys)
         {
             var stack = new Stack<IConfiguration>();
             stack.Push(configuration);
+            var rootSection = configuration as IConfigurationSection;
+            var prefixLength = (removePathFromChildKeys && rootSection != null) ? rootSection.Path.Length + 1 : 0;
```

Comment:
Note: this code is kind of ugly, if we represented the Root as simply having a null/empty path with the property on IConfiguration, we wouldn't have to always do this cast to IConfigurationSection every time we do traversal code on config looking for a Path...